### DISCUSSION TITLE
Performance Graph

### DISCUSSION THREAD

**Discussion by philipJFpfahl** [idaholab/moose#24009] [Q&A General] [answered]
Hello,

I want to do a runtime measurement of my main app and I am only concerned about the performance of the main app. Is it possible to measure only the main app runtime and memory usage, without the contribution of the sub-apps or do I have to change the main-sub structure of my application?

All the best
Philip

**Comment by GiudGiud** [comment, marked as answer]
Hello

Use the perf_graph in the main app and manually substract the multiapps portion from the total. 
You should be able to get the multiapps share in the main app, but if not, then use the perf graph in the subapp as well. 

Guillaume

**Reply by GiudGiud** [reply]
I'd try that. I've never done it so please let us know how it goes

**Reply by philipJFpfahl** [reply]
Okay, that was also my first idea.
So I have a main app and I am only interested in the time it takes to calculate one timestep within the main app. The sub-app only loads data and does mesh translation, which I do not want to measure.
So I look at the 
```
|                                   Section                                  | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |

|     MooseApp::execute                                                      |     1 |      0.000 |      0.000 |   0.00 |       0 |     42.512 |     42.512 |  98.74 |      38 |
|       MooseApp::executeExecutioner                                         |     1 |      0.007 |      0.007 |   0.02 |       1 |     42.512 |     42.512 |  98.74 |      38 |
|         Transient::PicardSolve                                             |     3 |      0.030 |      0.010 |   0.07 |       4 |     41.416 |     13.805 |  96.20 |      24 |

```
and divide it by the number of time steps: 41.416s / 4timestep ~ 10s/timestep. 
I assume that in this measurement the runtime of the ```Transient::PicardSolve``` of my sub-app is included, but not the time it takes to set up the sub-app. Then I can subtract 

```
sub-app0: |                                 Section                                | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |

sub-app0: |   Transient::PicardSolve                                               |     3 |      0.004 |      0.001 |   0.01 |       0 |      1.982 |      0.661 |   4.62 |      -5 |

```
Resulting in a runtime of ~ 41.416s / 4timestep-1.982s /4timestep ~ 9.5s/timestep.
Should I consider more of the execution time?

Philip

**Reply by GiudGiud** [reply]
This seems fine to ignore the setup tasks. 

If you are trying to draw conclusions from this, you ll like want to execute these timings on several independent runs btw
they can vary from one execution to the other

**Reply by philipJFpfahl** [reply]
Thank you. That helps a lot.
Yes this will need to be automated so that I can do several runs with several settings. Is there a Python tool for reading out the PerfGraf JSON file?
Philip

**Reply by GiudGiud** [reply]
@loganharbour any tips

We dont distribute one afaik, though i m sure there's something to read JSON files in python

**Reply by philipJFpfahl** [reply]
Okay, now I am confused :) 
I think that the  Performance Graph does some sort of double bookkeeping. Here are both the Performance Graph:
For the sub-app
```
sub0: Performance Graph:
sub0: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
sub0: |                                Section                               | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) 
sub0: | SealionTestApp (sub0)                                                |     1 |     42.861 |     42.861 |  93.55 |      17 |     45.815 |     45.815 | 100.00 |     111 |
sub0: |   Transient::PicardSolve                                             |     3 |      0.003 |      0.001 |   0.01 |       0 |      1.612 |      0.537 |   3.52 |      -5 |
```
And for the main-app
```
Performance Graph:
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                                 Section                                | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |
| SealionTestApp (main)                                                  |     1 |      0.005 |      0.005 |   0.01 |       2 |     45.975 |     45.975 | 100.00 |     137 |

|     Transient::PicardSolve                                             |     3 |      0.026 |      0.009 |   0.06 |       3 |     44.296 |     14.765 |  96.35 |      24 |

```
First of all my main app should be the heaviest part of the calculation. So that got me confused. Then I ran it by changing the number of nodes in the main-app while keeping the sub-app constant. 
That results in the Performance Graph of the sub-app appearing to "run" faster. 
```
sub0: Performance Graph:
sub0: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
sub0: |                                Section                               | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |
sub0: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------
sub0: | SealionTestApp (sub0)                                                |     1 |      0.367 |      0.367 |  11.63 |       2 |      3.154 |      3.154 | 100.00 |      97 |
```

So I assume that the performance graphs are not truly separated?


Philip

**Reply by GiudGiud** [reply]
This line is not
```
sub0: | SealionTestApp (sub0)                                                |     1 |     42.861 |     42.861 |  93.55 |      17 |     45.815 |     45.815 | 100.00 |     111 |
```
but this line is
```
sub0: |   Transient::PicardSolve                                             |     3 |      0.003 |      0.001 |   0.01 |       0 |      1.612 |      0.537 |   3.52 |      -5 |
```

the perf graph takes differences between start and end points. It s pretty simple. PicardSolve is solidly inside the MultiApp, however the first line of the graph is simply a counter started at the app initialization

**Reply by philipJFpfahl** [reply]
Ahh, I see.
So the main app Transient::PicardSolve tells me the total time spent on Transient::PicardSolve and the sub app Transient::PicardSolve tells me the time the sub app spends there. 

The first lines just tell me how long the main app needed to initialize the sub app.

Thank you
Philip

**Reply by GiudGiud** [reply]
First line is closer to the entire simulation time actually

And PicardSolve in the main app includes the subapp run time

**Reply by philipJFpfahl** [reply]
"First line is closer to the entire simulation time actually"
yes, I meant the difference between the two. 

"And PicardSolve in the main app includes the subapp run time" 
The PicardSolve subapp run time?

**Reply by GiudGiud** [reply]
> "And PicardSolve in the main app includes the subapp run time"
The PicardSolve subapp run time?

It should. The Picard in PicardSolve for the main app is actually for the picard iteration between the main app and the subapp

**Reply by philipJFpfahl** [reply]
yes that seems alright.

With a smaller mesh in the main app, the Picard solve gets smaller. But stays the same in the sub-app. Also, the solution seems alright.

thanks

**Reply by GiudGiud** [reply]
> With a smaller mesh in the main app, the Picard solve gets smaller. But stays the same in the sub-app. Also, the solution seems alright.

That seems correct. The subapp timing is not affected by mesh changes in the main app most likely.

**Reply by philipJFpfahl** [reply]
Yes I will run some tests and report back.

**Reply by philipJFpfahl** [reply]
Okay the results look odd to me. Maybe you have an idea why that is.

In the graph I plotted the runtime on y and the number of nodes nx * ny with nx = ny.
I think the "bumps" come from not enough runs, so there is some variance in the run times. 
But why does it follow a parabola?
I think it should follow a linear relation with the number of nodes.

![runtime](https://user-images.githubusercontent.com/95305404/230797400-0feada06-7567-4001-97df-247440bfbc9b.png)

ps: Tell me if I should a new discussion about that.

**Reply by GiudGiud** [reply]
Could be that the solve method is quadratic in the number of elements. What preconditioning are you using?

**Reply by philipJFpfahl** [reply]
So I was using NEWTON and BDP. But this happens for different preconditioning matrixes and different solvers.

**Reply by GiudGiud** [reply]
what is BDP?

**Reply by philipJFpfahl** [reply]
The default Block Diagonal Preconditioning. 
The transfer is included in the PicardSolve. Maybe that is interfering? 

Did you never experience this kind of runtime scaling?

**Reply by GiudGiud** [reply]
This scaling is typical if you are using direct solvers, eg with LU preconditioning. 

Transfers are typically fast. 
If you run this through a profiler (on macos, instruments is easy to use) you should see which part is driving this
https://mooseframework.inl.gov/application_development/profiling.html

**Reply by philipJFpfahl** [reply]
Okay. Thank you that was a great help.

Yes, I see the transfers for a deeper level. I will try out gperftool. 
Have a nice easter.

**Reply by GiudGiud** [reply]
so actually in the perfgraphoutput we can make it way more verbose
https://mooseframework.inl.gov/source/outputs/PerfGraphOutput.html

just set a level of 10

this should help even before profiling

**Reply by philipJFpfahl** [reply]
Yes I am running several times and then averaging over the Jsonoutput files. This is also this deep. And actually more than sufficient. 
The transfer of this problem is indeed super fast. It ranges at 0.5s even though I load all variables of this problem.

**Reply by GiudGiud** [reply]
ah ok. 
So in those outputs, do you see which section is taking long & scaling parabolically?

**Reply by philipJFpfahl** [reply]
Hello,

The section that scales quadratic is PicardSolve. Within it there are two functions that scale quadratic: Solve and computeNodalVars. Apparently MOOSE calculates the Nodes quadraticall in these functions.

**Reply by GiudGiud** [reply]
computeNodalVars should likely be linear with the size of the mesh (the mesh is not growing quadratically is it?)

Can you please paste your Executioner and Preconditioner blocks?

**Reply by philipJFpfahl** [reply]
I generate a square mesh like this:

```
[Mesh]
  type = GeneratedMesh
  allow_renumbering = false
  dim = 2
  xmin = 0
  xmax = 2
  nx = 100
  ymin = 0
  ymax = 2
  ny = 100
  parallel_type = replicated
[]
```
And change the values for ny and nx.
In the plot above, I calculate the number of nodes (shown on the x-axis) by calculating nx*ny.
```
[Executioner]
  type = Transient
  solve_type = NEWTON 
  num_steps = 3
  dt = 0.001
[]
```
```
[Preconditioning]
  [./smp]
    type = SMP
    full = true
    petsc_options = '-snes_converged_reason'
    petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -snes_max_it -sub_pc_factor_shift_type -pc_asm_overlap -snes_atol -snes_rtol '
    petsc_options_value = 'gmres asm lu 100 NONZERO 2 1E-14 1E-12'
  [../]
[]
```

**Reply by GiudGiud** [reply]
Can you try asm-ilu ?
I think asm-lu would still scale quadratically like lu, because the blocks treated by LU (roughly speaking) are still growing linearly

**Reply by philipJFpfahl** [reply]
So by changing
```
petsc_options_value = 'gmres asm ilu 100 NONZERO 2 1E-14 1E-12'
```
the scaling is still quadratic.

**Reply by GiudGiud** [reply]
Is this being run in serial or parallel? If in parallel how many ranks?

@oanaoana do you know the expected scaling of ASM with an ILU sub-pc ?

My next moves would be
- to go to the PETSc manual to understand what ASM and ILU are doing here precisely,
- to use an actual profiler to understand what section is growing quadratically

**Reply by philipJFpfahl** [reply]
It is serial.
I will add some more information to that.
```
MOOSE Version:           git commit ba019d8d56 on 2022-12-03
LibMesh Version:         
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Tue Apr 11 16:24:51 2023
Executable Timestamp:    Thu Feb 16 22:29:21 2023

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   5929
  Elems:                   5776
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                5929
  Num Local DOFs:          5929
  Variables:               "Dummy" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                118580
  Num Local DOFs:          118580
  Variables:               { "rho" "psi" "P" "a1" "a2" ... "C4" "C5" "C6" "C7" "C8" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP
```

**Reply by GiudGiud** [reply]
Your main variable is "dummy" ? Are you solving for it?

Is the concern the scaling of the nonlinear solve or of everything else handling all your variables?

**Reply by philipJFpfahl** [reply]
Yes it is like a crash test dummy. 
I only deal with AUX variables in this case. 
The quadratic scaling is in the Transient::PicardSolve.

I just ran a test and you were 100% correct. The problem is that I do not solve for a variable. And of course, the solver has no impact on the AUX system.
When I solve a problem with variables the lu solver scales linear and the ilu solver scales linear too.

I found the aux kernel, that causes the scaling. I calculate a matrix that is numberOfNodes squared. Thank you for your guidance in this problem!
Truly solved.

**Reply by philipJFpfahl** [reply]
Sorry if I wasted your time with a lack of information.
